osquery/tables/system/darwin/connected_displays.mm

```diff
@@ -0,0 +1,215 @@
+    // must be careful to properly clean up the memory. Any future modifications
+    // to this function should attempt to ensure there are no leaks.
+    CFURLRef bundle_url = CFURLCreateWithFileSystemPath(
+        kCFAllocatorDefault,
+        CFSTR("/System/Library/PrivateFrameworks/SPSupport.framework"),
+        kCFURLPOSIXPathStyle,
+        true);
+
+    if (bundle_url == nullptr) {
+      LOG(INFO) << "Error parsing SPSupport bundle URL";
+      return results;
+    }
+
+    CFBundleRef bundle = CFBundleCreate(kCFAllocatorDefault, bundle_url);
+    CFRelease(bundle_url);
+    if (bundle == nullptr) {
+      LOG(INFO) << "Error opening SPSupport bundle";
+      return results;
+    }
+
+    CFBundleLoadExecutable(bundle);
+
+    std::function<void()> cleanup = [&]() {
+      CFBundleUnloadExecutable(bundle);
+      CFRelease(bundle);
+    };
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
+
+    id cls = NSClassFromString(@"SPDocument");
+    if (cls == nullptr) {
+      LOG(INFO) << "Could not load SPDocument class";
+      cleanup();
+
+      return results;
+    }
+
+    SEL sel = @selector(new);
+    if (![cls respondsToSelector:sel]) {
+      LOG(INFO) << "SPDocument does not respond to new selector";
+      cleanup();
+
+      return results;
+    }
+
+    id document = [cls performSelector:sel];
+    if (document == nullptr) {
+      LOG(INFO) << "[SPDocument new] returned null";
+      cleanup();
+
+      return results;
+    }
+
+#pragma clang diagnostic pop
+
+    cleanup = [&]() {
+      CFRelease((__bridge CFTypeRef)document);
+      CFBundleUnloadExecutable(bundle);
+      CFRelease(bundle);
+    };
+
+    NSDictionary* report = [[[document reportForDataType:@"SPDisplaysDataType"]
+        objectForKey:@"_items"] lastObject];
+    NSArray* data = [report valueForKeyPath:@"spdisplays_ndrvs"];
+
+    for (NSString* obj in data) {
+      Row r;
+
+      if (data == nullptr) {
+        return results;
+      }
+
+      if ([obj valueForKey:@"_name"]) {
+        r["name"] = TEXT([[obj valueForKey:@"_name"] UTF8String]);
+      }
+
+      if ([obj valueForKey:@"_spdisplays_display-product-id"]) {
+        r["product_id"] = TEXT(
+            [[obj valueForKey:@"_spdisplays_display-product-id"] UTF8String]);
+      }
+
+      if ([obj valueForKey:@"_spdisplays_display-serial-number"]) {
+        r["serial_number"] = TEXT([[obj
+            valueForKey:@"_spdisplays_display-serial-number"] UTF8String]);
+      }
+
+      if ([obj valueForKey:@"_spdisplays_display-vendor-id"]) {
+        r["vendor_id"] = TEXT(
+            [[obj valueForKey:@"_spdisplays_display-vendor-id"] UTF8String]);
+      }
+
+      if ([obj valueForKey:@"_spdisplays_display-week"]) {
+        r["manufactured_week"] =
+            INTEGER([[obj valueForKey:@"_spdisplays_display-week"] intValue]);
+      } else {
+        r["manufactured_week"] = INTEGER(0);
```

Comment:
Are these "error states" where we cannot determine the value? iirc we are using `-1` for those